fix: fix various unit test failures in native_datafusion and native_iceberg_compat readers

parthchandra opened the pull request:
major changes :
- allow Uint64 to decimal and  FixedWidthBinary to Binary conversions in complex readers
- do not enable prefetch reads in tests if complex reader is enabled
- fix more incompatible checks in uint_8/uint_16 tests
- skip datetime rebase tests for complex readers (not supported)

There may be conflicts between this and #1413  (@mbutrovich) which removes the `cast_supported` method but can be reconciled afterwards

Without #1413 the failure counts are: 
```
native_datafusion: Tests: succeeded 658, failed 19, canceled 4, ignored 54, pending 0
native_iceberg_compat: Tests: succeeded 662, failed 15, canceled 4, ignored 54, pending 0
```

mbutrovich commented on spark/src/test/scala/org/apache/spark/sql/comet/ParquetDatetimeRebaseSuite.scala:
```diff
@@ -142,8 +145,13 @@ class ParquetDatetimeRebaseV1Suite extends ParquetDatetimeRebaseSuite {
-    super.test(testName, testTags: _*)(withSQLConf(SQLConf.USE_V1_SOURCE_LIST.key -> "") {
-      testFun
-    })(pos)
+    // Datasource V2 is not supported by complex readers so force the scan impl back
```
Let's be consistent and call them native readers. We will likely use these for more than just complex types at some point.

parthchandra replied:
changed the comment

mbutrovich commented:
#1413 shouldn't really conflict with this (I mean it will from a merge standpoint, but not from a logic standpoint). This just adds more true cases to `cast_supported` and #1413 removes the check entirely.

parthchandra commented:
> #1413 shouldn't really conflict with this (I mean it will from a merge standpoint, but not from a logic standpoint). This just adds more true cases to `cast_supported` and #1413 removes the check entirely.

I've updated the pr and removed the schema adapter changes. The number of failures has gone up but they should be taken care of by #1413

parthchandra commented:
@kazuyukitanimura @andygrove @huaxingao @comphead review requested.

andygrove reviewed (approved):
LGTM. Thanks @parthchandra

comphead commented on common/src/main/java/org/apache/comet/parquet/NativeBatchReader.java:
```diff
@@ -260,7 +260,8 @@ public void init() throws URISyntaxException, IOException {
 
     String timeZoneId = conf.get("spark.sql.session.timeZone");
-    Schema arrowSchema = Utils$.MODULE$.toArrowSchema(sparkSchema, timeZoneId);
+    // Native code uses "UTC" always as the timeZoneId when converting from spark to arrow schema.
```
👍

comphead commented on common/src/main/java/org/apache/comet/parquet/NativeBatchReader.java:
```diff
@@ -260,7 +260,8 @@ public void init() throws URISyntaxException, IOException {
     } ////// End get requested schema
 
     String timeZoneId = conf.get("spark.sql.session.timeZone");
```
wondering should we always use UTC then and remove reading the tz from spark parameter?

comphead reviewed (approved):
lgtm thanks @parthchandra

kazuyukitanimura commented on spark/src/test/scala/org/apache/comet/exec/CometColumnarShuffleSuite.scala:
```diff
@@ -740,28 +740,48 @@ abstract class CometColumnarShuffleSuite extends CometTestBase with AdaptiveSpar
-          $"_20").foreach { col =>
+        val fieldsToTest = if (CometSparkSessionExtensions.usingDataFusionParquetExec(conf)) {
+          Seq(
+            $"_1",
+            $"_4",
+            $"_5",
+            $"_6",
+            $"_7",
+            $"_8",
+            $"_11",
+            $"_12",
+            $"_13",
+            $"_14",
+            $"_15",
+            $"_16",
+            $"_17",
+            $"_18",
+            $"_19",
+            $"_20")
```
We are skipping 2, 3, 9, and 10, I guess it is due to unsupported uints
Perhaps we can put TODO

parthchandra replied:
We are skipping ByteType and ShortType (both signed and unsigned). I doubt if the TODO will get addressed (it appears that this is unlikely to be addressed in arrow-rs), but I've added it.

kazuyukitanimura commented on spark/src/test/scala/org/apache/comet/parquet/ParquetReadSuite.scala:
```diff
@@ -97,13 +102,19 @@ abstract class ParquetReadSuite extends CometTestBase {
-          StructField("f2", StringType))) -> false,
+          StructField("f2", StringType))) -> usingNativeIcebergCompat,
       MapType(keyType = LongType, valueType = DateType) -> false,
-      StructType(Seq(StructField("f1", ByteType), StructField("f2", StringType))) -> false,
+      StructType(
+        Seq(
+          StructField("f1", ByteType),
+          StructField("f2", StringType))) -> usingNativeIcebergCompat,
       MapType(keyType = IntegerType, valueType = BinaryType) -> false).foreach {
```
Just checking, looks like `MapType` is not supported yet?

parthchandra replied:
Correct

kazuyukitanimura commented on spark/src/test/scala/org/apache/comet/parquet/ParquetReadSuite.scala:
```diff
@@ -97,13 +102,19 @@ abstract class ParquetReadSuite extends CometTestBase {
       case (dt, expected) =>
         assert(CometScanExec.isTypeSupported(dt) == expected)
-        assert(CometBatchScanExec.isTypeSupported(dt) == expected)
+        // usingDataFusionParquetExec does not support CometBatchScanExec yet
+        if (!usingDataFusionParquetExec(conf)) {
```
> usingDataFusionParquetExec does not support CometBatchScanExec yet

I assume this means we are not supporting V2 yet. But
`CometScanExec.isTypeSupported` == `CometBatchScanExec.isTypeSupported` so I guess we can still pass this test regardless `usingDataFusionParquetExec`?

kazuyukitanimura commented on spark/src/test/scala/org/apache/comet/parquet/ParquetReadSuite.scala:
```diff
@@ -1001,7 +1012,7 @@ abstract class ParquetReadSuite extends CometTestBase {
 
             withParquetDataFrame(data, schema = Some(readSchema)) { df =>
-              if (enableSchemaEvolution) {
+              if (enableSchemaEvolution || usingDataFusionParquetExec(conf)) {
```
Hmm does this mean that `usingDataFusionParquetExec==true` we always use schema evolution regardless `COMET_SCHEMA_EVOLUTION_ENABLED` This means it will be incompatible with Spark 3.x?

kazuyukitanimura replied:
I feel we need a TODO comment or an issue ticket to track this.

parthchandra replied:
done

parthchandra commented:
@kazuyukitanimura addressed your last comment and also rebased (there were merge conflicts).
Test failure count after rebase: 
```
native_datafusion:     Tests: succeeded 661, failed 14, canceled 6, ignored 54, pending 0
native_iceberg_compat: Tests: succeeded 663, failed 12, canceled 6, ignored 54, pending 0
```

kazuyukitanimura commented:
merged thank you @parthchandra @comphead @andygrove @mbutrovich
